Feat: openapi py client

lyuyangh opened the pull request:
1. use [openapi generator](https://github.com/OpenAPITools/openapi-generator) instead of [swagger codegen](https://github.com/swagger-api/swagger-codegen) to generate python client because it offers much better documentation and more features
    - I think it may be a good idea to migrate to openapi generator for flask and ts too. This migration shouldn't be too bad. The templates and the CLI commands are very similar.
2. I made some small modifications to the python client template so the template is included in the PR.
3. openapi generator provides two options for python clients: `python` and `python-experimental`. `python-experimental` provides better typing but it is complicated to use and to work with. I tried to fix the bugs in its template but fixing one bug reveals more bugs 🥲 and I don't think it's an easy fix. So, currently, I'm using the `python` template. There's limited type hint but it works great and it's relatively easy to figure out the input and outputs from the docstrings.
4. The generated python client is released as a wheel in this repo. We can release it to PyPI in the future if it makes things easier. It uses the commit SHA as the version number. We may need to figure out a better way to version the generated client.
5. https://github.com/neulab/explainaboard_cli is a wrapper for the generated client to make it easier to use.

neubig reviewed (commented):
Thanks so much for this @lyuyangh ! Because the code is mostly auto-generated this looks fine, but it'd be really nice to have a README somewhere with a brief explanation of the process to generate this API code, including:
(1) what is the overall structure of the code (you could probably copy-paste some of the info from the this PR)
(2) when/how will we need to update the code in this repo?
(3) when/how will we need to update the thin CLI client?

Thanks a bunch again!

lyuyangh commented:
Sounds good! I'll also make some adjustments to how we version the python client as we discussed last time. I can probably finish these by tomorrow.

lyuyangh commented:
Hi @neubig ! Thanks! I updated these things since you commented on the PR:
1. `explainaboard_api_client` and `explainaboard_cli` are both released to PyPI now
2. the version number of `explainaboard_api_client` is determined by `info.version` in openapi.yaml. If openapi.yaml is changed but the version is not updated, the release pipeline will complain.
3. added a section in the README to explain how the python client is generated.
4. added instructions about when&how to update the packages in the README in `explainaboard_cli`
5. GET /info now returns the API version that's running on the server so the users know which `explaina_api_client` package they should use. We can display this info on the UI in the future.

Please let me know if there's anything else you would like me to add! Thanks!

neubig commented:
Thanks a lot @lyuyangh ! I was looking through this and I think it basically looks good. There are a few small things that we might be able to do to make the thin client even easier to use, but I think they can be saved for a future PR.

One thing: on thinking about it, `explainaboard_cli` seems like a misnomer though. It's not a command-line interface, but rather a client. Maybe we could just name the repo (and the pypi package) `explainaboard_client`. The gives a bit of overlap with `explainaboard_api_client` but perhaps we could also change that to something like `explainaboard_api` or something? Sorry I just thought about it now, but I realized this when I was looking at the tests in the thin client's repository.

lyuyangh commented:
@neubig Yeah, that makes sense. I have been calling it a CLI because I think we can turn it into an actual CLI in the future. e.g. Users can submit system results in the terminal and maybe view and scroll (w/ arrow keys) through all the datasets/systems/tasks without going to the web UI (or maybe they don't have a GUI so that's the only option). This is similar to AWS CLI. I think at least we should provide a command for the user to log in once and save the API key in a file so they don't have to hardcode the API key in the code. Do you think this would be useful? If you think we most likely don't need to support this in the near future, I'll change the package name to `explainaboard_api`. I should have explained this earlier. Sorry for the confusion!

neubig commented:
Could we call it `explainaboard_client` which encompasses both CLI and API interfaces? I do think it'd be nice to provide both.

lyuyangh commented:
Sounds good! I renamed it `explainaboard_client`. And yeah, I agree it'll make things easier for our users if the two are bundled together. I'll think more about it to see if there's a good way to do that.

neubig reviewed (approved):
Thanks a bunch @lyuyangh , this is great. I'm happy to merge now.

In terms of the next steps for me to test that this is working, we can merge this PR then I could try to run the client with "local" or "staging" settings?

lyuyangh commented:
Thanks a lot! Yeah, once this PR is deployed, the tests I included in the client repo should work for all three environments.

lyuyangh commented:
The release action fails because I have already released 0.1.0 when I was testing. It should work when we bump the API spec (openapi.yaml) version next time.

latest versions for:
- `explainaboard_api_client`==0.1.0
- `explainaboard_client`==0.0.3
